traffic_ops/traffic_ops_golang/server/servers_server_capability.go

```diff
@@ -453,77 +443,178 @@ func AssignMultipleServerCapabilities(w http.ResponseWriter, r *http.Request) {
 	if err != nil {
-		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, err)
+		useErr, sysErr, statusCode := api.ParseDBError(err)
+		api.HandleErr(w, r, tx, statusCode, useErr, sysErr)
 		return
 	}
 
-	userErr, sysErr, errCode = dbhelpers.CheckIfCurrentUserCanModifyCDN(tx, string(cdnName), inf.User.UserName)
+	var alerts tc.Alerts
+	if mssc.PageType == "sc" {
+		alerts = tc.CreateAlerts(tc.SuccessLevel, "Assign Server(s) to a capability")
+	} else {
+		alerts = tc.CreateAlerts(tc.SuccessLevel, "Assign Server Capability(ies) to a server")
+	}
+	api.WriteAlertsObj(w, r, http.StatusOK, alerts, mssc)
+	return
+}
+
+// DeleteMultipleServersCapabilities deletes multiple servers to a capability or multiple server capabilities to a server
+func DeleteMultipleServersCapabilities(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	tx := inf.Tx.Tx
 	if userErr != nil || sysErr != nil {
-		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
 		return
 	}
+	defer inf.Close()
 
-	//Delete existing rows from server_server_capability for a given server
-	_, err = tx.Exec("DELETE FROM server_server_capability ssc WHERE ssc.server=$1", msc.ServerID)
-	if err != nil {
-		useErr, sysErr, statusCode := api.ParseDBError(err)
-		api.HandleErr(w, r, tx, statusCode, useErr, sysErr)
+	var mssc tc.MultipleServersCapabilities
+	if err := json.NewDecoder(r.Body).Decode(&mssc); err != nil {
+		api.HandleErr(w, r, tx, http.StatusBadRequest, fmt.Errorf("error decoding DELETE request body into MultipleServersCapabilities struct %w", err), nil)
 		return
 	}
 
-	multipleServerCapabilities := make([]string, 0, len(msc.ServerCapabilities))
+	if len(mssc.ServerIDs) >= 1 {
+		errCode, userErr, sysErr = checkExistingServer(tx, mssc.ServerIDs, inf.User.UserName)
+		if userErr != nil || sysErr != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+			return
+		}
+	}
 
-	mscQuery := `WITH inserted AS (
-		INSERT INTO server_server_capability
-		SELECT "server_capability", $2
-		FROM UNNEST($1::text[]) AS tmp("server_capability")
-		RETURNING server_capability
-		)
-		SELECT ARRAY_AGG(server_capability)
-		FROM (
-			SELECT server_capability
-			FROM inserted
-		) AS returned(server_capability)`
+	//Delete existing rows from server_server_capability for a given server or for a given capability
+	const delQuery = `DELETE FROM server_server_capability ssc WHERE `
+	var dq string
+	var alerts tc.Alerts
+	if mssc.PageType == "sc" {
+		dq = delQuery + fmt.Sprintf("ssc.server_capability='%v'", mssc.ServerCapabilities[0])
+		if len(mssc.ServerIDs) == 1 {
+			dq = dq + fmt.Sprintf(" AND ssc.server=%v", mssc.ServerIDs[0])
+		}
+	} else {
+		dq = delQuery + fmt.Sprintf("ssc.server=%v", mssc.ServerIDs[0])
+		if len(mssc.ServerCapabilities) == 1 {
+			dq = dq + fmt.Sprintf(" AND ssc.server_capability='%v'", mssc.ServerCapabilities[0])
```

Comment:
Actually, just wanted to point out something that CodeQL found, since Srijeet hasn't mentioned it.

You should never ever ever use `fmt.Sprintf` to build queries from user data like this. You MUST use prepared database statements. The way this is right now, I can submit a request like:

```json
{"serverIds": [2], "serverCapabilities": ["test'; DROP TABLE tm_user CASCADE; --"]}
```

(to DELETE /api/4.1/multiple_servers_capabilities)

and I'll get a response like
```json
{
	"response": {
		"serverCapabilities": [
			"test'; DROP TABLE tm_user CASCADE; --"
		],
		"serverIds": [
			2
		],
		"pageType": ""
	}
}
```
... and it has now dropped the users, changelogs, and Delivery Service Request tables from the database. All data lost, API rendered completely unusable because nobody can log in as any user.

Reply:
I know that.. it was part of my next update but thanks for an explanation.